packages/block-library/src/site-logo/edit.js

```diff
@@ -349,14 +401,59 @@ export default function LogoEdit( {
 				alt: mediaItem.alt_text,
 			},
 			isRequestingMediaItem: _isRequestingMediaItem,
+			siteIconId: _siteIconId,
 		};
 	}, [] );
 
+	const { getGlobalBlockCount } = useSelect( blockEditorStore );
 	const { editEntityRecord } = useDispatch( coreStore );
-	const setLogo = ( newValue ) =>
+
+	useEffect( () => {
```

Comment:
Hi there!

The logic here doesn't make sense and should be, it's creating bugs like #37760

Basically the logic here assumes than when we "unmount" the site logo block, we're removing it when it's not always the case, unmounting can happen in several ways:

 - Switching to code editor in post editor
 - Changing the current view (template/template part) in the site editor.

While I'm not sure about the rest of this PR, this effect should be removed.

Reply:
@stacimc I removed the hook and went through the testing instructions above under **"Test discarding unsaved changes:"**

I couldn't find any weird side effects, that is, the editor recognised and kept my updates when viewing previews and when switching between code/visual modes. 

Is there something I'm not testing right?

Reply:
The logic here doesn't assume that unmounting the block means we're removing it. It runs every time the block is unmounted, but it then [checks the global block count](https://github.com/WordPress/gutenberg/pull/35892/files#diff-0d5debfeb0c6fe28a3979c43ecb9029bae1617d404b1c92f506ba754333f9647R432) to determine whether the Site Logo blocks have actually been removed; otherwise it should do nothing.

I think the problem is in the Site Editor, that `getGlobalBlockCount` isn't behaving the way I believed it would. Although all my own test scenarios still work for me, I'm able to reproduce the #37760 bug by:
* Add a Site Logo block and make a change in the Header part
* Add a different block to a different template part which does _not_ contain a Site Logo block
* Hit Undo and see that the 'Redo' button is disabled

When the Site logo block unmounts, `getGlobalBlockCount` returns 0 this time.

Reply:
I'll put up a PR to remove the logic for now and then start looking into a better way of doing this. I do think it would be nice to discard those unsaved changes when the block is removed, if I can get it working.

Reply:
#37895 fixes the regression, and #37900 has been opened to investigate adding the functionality back in safely.

Reply:
Thanks @stacimc `getGlobalBlockCount` is not reliable because the `core/block-editor` store is more like an internal store to an instance of the block editor and don't say anything about the presence of a block in a template (or all templates). It's just specific to the current state of the screen.